### DISCUSSION TITLE
Can we create a new user for exists bucket with rook?

### DISCUSSION THREAD

**Discussion by zhucan** [rook/rook#10365] [Q&A]
for example use “object-user.yaml”

@travisn

**Comment by zhucan** [comment]
for example: Create users through rook using the CephObjectStoreUser CRD and giving them all capabilities [each user still can not view each others buckets @travisn

**Reply by thotz** [reply]
what are the capabilities which you are looking for, can you please be more specific?

**Reply by zhucan** [reply]
We want to create a new user that can access different buckets.

**Reply by thotz** [reply]
Please see the below comment, u may need to add a bucket policy to all the buckets that need access. This is explicitly defined for the s3 protocol. AFAIR there is also user policy which s3 supports but does not implement in ceph. If you want to just see the all buckets and their stats(not the contents) via a specific user like an admin then u can use [admin ops API](https://docs.ceph.com/en/latest/radosgw/adminops/). It is similar to `radosgw-admin` commands but via rest APIs.

**Reply by zhucan** [reply]
we want to create different buckets and with same owner. not one user to per bucket. In the product, we  want to access different buckt through same ak  and sk (use same  user). @thotz

**Reply by thotz** [reply]
I guess bucketpolicy is the best approach for you folks. You can also check with RGW devs what is best to way to implement it with minimal steps

**Reply by zhucan** [reply]
ok

**Comment by richardelling** [comment]
The mechanism is bucket policies. It works and you can edit them in the data path as expected with many S3 clients.

Having recently gone through trying to do something similar to this proposal, the engineering problem is quite difficult to solve for the general case. Some major obstacles are:

1. Since clients can edit the bucket policies in many unpredictable ways and you cannot (or should not) prevent this. This makes the reconciliation task very complicated: how can you ensure the capabilities expected are allowed and ensure an end user didn't create a denial that undoes what you expect. And then it gets more complicated by the resource specification and action details. For example, there is a set of permitted actions to the bucket resource and another for the bucket contents.
2. Deleting a user gets complicated because of the use of globs.
3. Going from a user-generated set of bucket policies to something higher level that represents intent (eg, share this bucket with upload-only permissions to a user) is difficult.
4. What should be done to bucket policies when a user is deleted?

So you end up having to do a very specific pattern of bucket policies that might or might not meet your customer's requirements, but the customer can always edit the policies to get what they want. One approach taken by some clients (eg minio client) is to have a public user policy template. But this only works for public users and the default policies of upload, download, or both. Again, this simplifies the engineering work, but is also quite constrained.

I don't see a generic solution to these problems that makes sense to put into a rook CRD. Hence, I agree completely with @thotz comment that you need to be very specific in your request for capabilities.
